rasterio/_vsiopener.pyx

```diff
@@ -176,24 +299,87 @@ cdef int pyopener_close(void *pFile) except -1 with gil:
 
 
 @contextlib.contextmanager
-def _opener_registration(urlpath, mode, opener):
+def _opener_registration(urlpath, mode, obj):
     registry = _OPENER_REGISTRY.get()
-    if (urlpath, mode) in registry:
-        if registry[(urlpath, mode)] != opener:
-            raise OpenerRegistrationError(f"Opener already registered for urlpath and mode")
+    from urllib.parse import urlparse
+    parsed_uri = urlparse(urlpath)
+    path_to_check = Path(parsed_uri.path)
+    parent = path_to_check.parent
+
+    key = ((parsed_uri.scheme, parsed_uri.netloc, parent.as_posix()), mode[0])
+    # Might raise.
+    opener = _create_opener(obj)
+
+    if key in registry:
+        if registry[key] != opener:
+            raise OpenerRegistrationError(f"Opener already registered for urlpath and mode.")
         else:
             try:
                 yield f"{PREFIX}{urlpath}"
             finally:
                 registry = _OPENER_REGISTRY.get()
-                _ = registry.pop((urlpath, mode), None)
+                _ = registry.pop(key, None)
                 _OPENER_REGISTRY.set(registry)
     else:
-        registry[(urlpath, mode)] = opener
+        registry[key] = opener
         _OPENER_REGISTRY.set(registry)
         try:
             yield f"{PREFIX}{urlpath}"
         finally:
             registry = _OPENER_REGISTRY.get()
-            _ = registry.pop((urlpath, mode), None)
+            _ = registry.pop(key, None)
             _OPENER_REGISTRY.set(registry)
+
+
+class _FileOpener:
+    """Adapts a Python file object to the opener interface."""
+    def __init__(self, obj):
+        self._obj = obj
+    def open(self, path, mode="r", **kwds):
+        return self._obj(path, mode=mode, **kwds)
+    def is_file(self, path):
+        return True
+    def is_dir(self, path):
+        return False
+    def ls(self, path):
+        return []
+    def size(self, path):
+        with self._obj(path) as f:
+            f.seek(0, os.SEEK_END)
+            return f.tell()
+
+
+class _FilesystemOpener:
+    """Adapts an fsspec filesystem object to the opener interface."""
+    def __init__(self, obj):
+        self._obj = obj
+    def open(self, path, mode="r", **kwds):
+        return self._obj.open(path, mode=mode, **kwds)
+    def is_file(self, path):
+        return self._obj.is_file(path)
+    def is_dir(self, path):
+        return self._obj.is_dir(path)
+    def ls(self, path):
+        return self._obj.ls(path)
+    def size(self, path):
+        return self.size(path)
```

Comment:
@sgillies I'm not sure to get all the things in this *draft* PR, but quick question about this `_FilesystemOpener`, do you expect the opener object to have ll of those methods (is_file, is_dir and ls) ?

Reply:
@vincentsarago yes. `is_file()/is_dir()` are required to support the stat callback. `ls()` is required to support the read_dir callback. I'm basing the interface on fsspec.

Reply:
Those are all the methods that are needed. I don't intend to support making directories, copying files, renaming files, or deleting files.